### PR TITLE
[Catchup] Emit Attributable Action Events

### LINKED ISSUE #2785: [CATCHUP] - Add an event emitted on every attributable action
### What is this task and why do we need to work on it?

When we restart, we need to pick a view number that is higher than any view in which we have previously taken an attributable action (vote, propose, etc). Otherwise, it is possible we will take conflicting actions in the same view before and after the restart, and an observer who is collecting evidence of both actions could slash us (not to mention this might lead to a fork).

HotShot should emit an event `AttributableAction { view_number }` any time such an action is taken, so that the application can keep track of the highest such view number and ensure we restart in a later view (the ability of the application to control the view number in which we restart is already provided by `HotShotInitializer`).

Note that if we are conservative and every node restarts in a later view or even a different view, this is fine: view sync will take care of it.

### What work will need to be done to complete this task?

Define the event, emit it in all the places where we do something attributable.

### Are there any other details to include?

Note that there is still a highly unlikely race condition: if we shut down in the very short window between taking such an action  and processing the corresponding event, we may restart in the wrong view. In the short term, we may just want to be conservative and emit the event _before_ taking the action. Long term, we should integrate the application storage layer more tightly into consensus logic, so that consensus must warn the application of the intent to take an attributable action _and_ get confirmation that the view number was stored successfully, before actually taking the action.

### What are the acceptance criteria to close this issue?

See above

### Branch work will be merged to (if not the default branch)

_No response_

### PR DESCRIPTION
**Pull request by bfish713**
Closes #2785 

### This PR: 

Emits an event after we send any attributable message to the network.  This is for the application to be able to restart with view that is higher than the previous attributable action view.  

### This PR does not: 

Doesn't handle the case where something goes wrong after a messages hits the wire, but before the event can be sent to or handled by the application

### Key places to review: 

Changes to `NetworkEventTaskState`

### CODE REVIEW

**Review by jbearer — COMMENTED** [review]
This looks good. But I'm wondering if we should just go ahead and integrate this into the `Storage` API that @jparr721 has created, so that we can actually prevent the node from taking an action if we're unable to store the view? I didn't expect that to be ready so soon but given that it is...

That would just look like adding a `record_action(action, view)` method to the storage trait and calling it n all the places where we currently set `maybe_action`

**Comment by bfish713** [comment]
> This looks good. But I'm wondering if we should just go ahead and integrate this into the `Storage` API that @jparr721 has created, so that we can actually prevent the node from taking an action if we're unable to store the view? I didn't expect that to be ready so soon but given that it is...
> 
> That would just look like adding a `record_action(action, view)` method to the storage trait and calling it n all the places where we currently set `maybe_action`

Yes I agree with this, I was thinking we'd also add the recording of high_qc to this storage as well.  I'll update this after the storage changes are merged

**Review by jbearer — COMMENTED** [review]
Action storage looks good. Are we also going to add high QC storage and fix the `HotShotInitializer` handling of the high QC? Or is that for later?

**Comment by bfish713** [comment]
> Action storage looks good. Are we also going to add high QC storage and fix the `HotShotInitializer` handling of the high QC? Or is that for later?

I have a different PR I'm working on for that

File: crates/example-types/src/storage_types.rs [hunk 1]
```diff
@@ -64,4 +63,19 @@ impl<TYPES: NodeType> Storage<TYPES> for TestStorage<TYPES> {
...
+    ) -> Result<()> {
+        Ok(())
+    }
+
+    async fn update_high_qc(
+        &self,
+        _qc: hotshot_types::simple_certificate::QuorumCertificate<TYPES>,
+    ) -> Result<()> {
+        Ok(())
+    }
```
Comment (jparr721) [inline crates/example-types/src/storage_types.rs:80]: Should these have some sort of dummy implementation?

Comment (bfish713) [reply crates/example-types/src/storage_types.rs:80]: They could, but there isn't a way to even read from this so I'm not sure how valuable that would be.

File: crates/task-impls/src/network.rs [hunk 2]
```diff
@@ -377,7 +406,23 @@ impl<TYPES: NodeType, COMMCHANNEL: ConnectedNetwork<Message<TYPES>, TYPES::Signa
         let view = message.kind.get_view_number();
         let committee = membership.get_whole_committee(view);
         let net = self.channel.clone();
+        let storage = self.storage.clone();
         async_spawn(async move {
+            if let Some(action) = maybe_action {
```
[the remark this answers is not in the document]
Comment (bfish713) [reply crates/task-impls/src/network.rs:411]: It is probably, but we don't have a working network task test (it's ignored and would need to be completely rewritten).  I can do that in another PR.  We have an issue for it: https://github.com/EspressoSystems/HotShot/issues/2357

File: crates/task-impls/src/network.rs [hunk 3]
```diff
@@ -241,64 +252,82 @@ impl<TYPES: NodeType, COMMCHANNEL: ConnectedNetwork<Message<TYPES>, TYPES::Signa
     /// Panic sif a direct message event is received with no recipient
     #[allow(clippy::too_many_lines)] // TODO https://github.com/EspressoSystems/HotShot/issues/1704
     #[instrument(skip_all, fields(view = *self.view), name = "Network Task", level = "error")]
-
     pub async fn handle_event(
         &mut self,
         event: Arc<HotShotEvent<TYPES>>,
         membership: &TYPES::Membership,
     ) -> Option<HotShotTaskCompleted> {
+        let mut maybe_action = None;
```
[the remark this answers is not in the document]
Comment (bfish713) [reply crates/task-impls/src/network.rs:260]: I just didn't want keep growing the tuple for a thing that only applies to a few of the events.  Not sure what's better

**Review by jparr721 — APPROVED** [review]

**Pull request merged**